qiskit/ignis/mitigation/measurement/filters.py

```diff
@@ -176,5 +182,202 @@ def fun(x):
+                    is an integer (e.g. for use with the tomography data)
+                Form4: a qiskit Result
+
+            method (str): fitting method. If None, then least_squares is used.
+                'pseudo_inverse': direct inversion of the cal matrices
+                'least_squares': constrained to have physical probabilities
+
+        Returns:
+            The corrected data in the same form as raw_data
+        """
+
+        all_states = count_keys(self.nqubits)
+        num_of_states = 2**self.nqubits
+
+        # check forms of raw_data
+        if isinstance(raw_data, dict):
+            # counts dictionary
+            data_format = 0
+            # convert to form2
+            raw_data2 = [np.zeros(num_of_states, dtype=float)]
+            for state, count in raw_data.items():
+                stateidx = int(state, 2)
+                raw_data2[0][stateidx] = count
+
+        elif isinstance(raw_data, list):
+            size_ratio = len(raw_data)/num_of_states
+            if len(raw_data) == num_of_states:
+                data_format = 1
+                raw_data2 = [deepcopy(raw_data)]
+            elif int(size_ratio) == size_ratio:
+                data_format = 2
+                size_ratio = int(size_ratio)
+                # make the list into chunks the size of num_of_states
+                # for easier processing
+                raw_data2 = np.zeros([size_ratio, num_of_states])
+                for i in range(size_ratio):
+                    raw_data2[i][:] = raw_data[
+                        i * num_of_states:(i + 1)*num_of_states]
+            else:
+                raise QiskitError("Data list is not an integer multiple "
+                                  "of the number of calibrated states")
+
+        elif isinstance(raw_data, qiskit.result.result.Result):
+
+            # extract out all the counts, re-call the function with the
+            # counts and push back into the new result
+            new_result = deepcopy(raw_data)
+
+            for resultidx, _ in enumerate(raw_data.results):
+                new_counts = self.apply(
+                    raw_data.get_counts(resultidx), method=method)
+                new_result.results[resultidx].data.counts = \
+                    new_result.results[resultidx]. \
+                    data.counts.from_dict(new_counts)
+
+            return new_result
+
+        else:
+            raise QiskitError("Unrecognized type for raw_data.")
+
+        if method == 'pseudo_inverse':
+            pinv_cal_matrices = []
+            for cal_mat in self._cal_matrices:
+                pinv_cal_matrices.append(la.pinv(cal_mat))
+
+        # Apply the correction
+        for data_idx, _ in enumerate(raw_data2):
+
+            if method == 'pseudo_inverse':
+                inv_mat_dot_raw = np.zeros([num_of_states], dtype=float)
+                for state1_idx, state1 in enumerate(all_states):
+                    for state2_idx, state2 in enumerate(all_states):
```

Comment:
This only needs to be over indices with non-zero counts?

Reply:
Fixed